### PR TITLE
Feature/dat 604 gestion des scopes incompatibles entre eux

### PR DESCRIPTION
**Pull request by Isalafont**
- [x] Reste à faire des test Cucumber

**_Restera encore une validation à gérer mais elle est intrinsèquement liée à l'expression de besoins spécifique qui n'est pas encore intégré ici._**

Le Front va être géré dans une autre PR

### CODE REVIEW

File: app/models/authorization_request/api_impot_particulier.rb [hunk 1]
```diff
@@ -22,4 +48,28 @@ class AuthorizationRequest::APIImpotParticulier < AuthorizationRequest
...
+    return if scopes.intersect?(MANDATORY_REVENUE_YEARS)
+
+    errors.add(:scopes, :invalid, message: 'sont invalides : Vous devez cocher au moins une année de revenus souhaitée avant de continuer')
+  end
+
+  def validate_exclusive_years_scope_combination
+    return unless scope_exists_in_each_arrays?(*EXCLUSIVE_REVENUE_YEARS)
+
+    errors.add(:scopes, :invalid, message: "sont invalides : Vous ne pouvez pas sélectionner la donnée 'avant dernière année de revenu, si la dernière année de revenu est indisponible' avec d'autres années de revenus")
+  end
```
Comment (skelz0r) [inline app/models/authorization_request/api_impot_particulier.rb:64]: Le code est un poil complexe, et le nom un poil trop générique.

Ici ce qu'on test est que si `dgfip_annee_n_moins_2_si_indispo_n_moins_1` est sélectionné on ne peut pas sélectionner les 3 autres.

D'un point de vue code ceci me paraît plus simple à comprendre:

```ruby
return unless scopes.include?('dgfip_annee_n_moins_2_si_indispo_n_moins_1')
return unless scopes.intersect?(%w[dgfip_annee_n_moins_1 dgfip_annee_n_moins_2 dgfip_annee_n_moins_3])
```

après c'est kiff/kiff, imo si on fait un effort sur le nom ça sera peut-être mieux.

Il faudrait peut-être trouver un nom qui traduit la réalité de cette règle métier, j'arrive pas à mettre le doigt dessus mais si tu as ça serait cool.

Ce n'est pas très grave, c'est surtout de la réflexion.

Le seul point que j'aimerais qu'on corrige ici est de retirer le `validate_` dans le nom, donc si on garde l'existant ça ferait un truc du genre: `validate :years_scope_are_compatible`

File: app/models/authorization_request/api_impot_particulier.rb [hunk 2]
```diff
@@ -22,4 +48,28 @@ class AuthorizationRequest::APIImpotParticulier < AuthorizationRequest
...
+
+  def validate_incompatible_scopes
```
Comment (skelz0r) [inline app/models/authorization_request/api_impot_particulier.rb:66]: Bon là clairement c'est de la spec obscure, je n'ai aucune idée de si c'est valide ou non.. quelle est la justification métier de la DGFIP ici ? Parce que ça serait bien de le traduire dans le code (au même titre que la méthode du dessus).

Après ce n'est pas bloquant, juste techniquement il n'y a rien à dire, car il n'y a aucune justification métier donc on peut mettre n'importe quoi dedans :D

File: spec/models/authorization_request/api_impot_particulier_spec.rb [hunk 3]
```diff
@@ -45,4 +49,72 @@
...
+  describe 'validation of exclusive revenue years scope combination' do
+    before { authorization_request.current_build_step = 'scopes' }
+
+    AuthorizationRequest::APIImpotParticulier::EXCLUSIVE_REVENUE_YEARS.each do |exclusive_group|
+      context "with valid exclusive years scope combination: #{exclusive_group.join(', ')}" do
+        let(:scopes) { exclusive_group }
+
+        it { is_expected.to be_valid }
+      end
+    end
```
Comment (skelz0r) [inline spec/models/authorization_request/api_impot_particulier_spec.rb:80]: clairement ici tu peux simplifier en en testant 1 seul ça sera plus clair

File: spec/models/authorization_request/api_impot_particulier_spec.rb [hunk 4]
```diff
@@ -45,4 +49,72 @@
...
+      let(:scopes) { %w[dgfip_annee_df_au_3112_si_deces_ctb_mp dgfip_annee_n_moins_2_si_indispo_n_moins_1] }
+
+      it { is_expected.to be_valid }
+    end
+
+    context 'with invalid complex incompatible scopes combination from array_1 and array_2' do
+      let(:scopes) { %w[dgfip_annee_df_au_3112_si_deces_ctb_mp dgfip_annee_n_moins_2_si_indispo_n_moins_1 dgfip_indiIFI dgfip_RevDecl_Cat1_Tspr] }
+
+      it { is_expected.not_to be_valid }
+    end
```
Comment (skelz0r) [inline spec/models/authorization_request/api_impot_particulier_spec.rb:112]: je pense qu'ici c'est pareil, tu testes 1 cas évident, 1 cas complexe, 1 cas valide et pas besoin de plus imo

File: features/habilitations/dgfip/api_impots_particuliers_editeur.feature [hunk 5]
```diff
@@ -15,6 +15,31 @@ Fonctionnalité: Soumission d'une demande d'habilitation API Impôts Particulier
     * je renseigne le cadre légal
     * je clique sur "Suivant"
 
+  Scénario: Je soumets une demande d'habilitation sans scopes d'années de revenue cochés.
```
Comment (skelz0r) [inline features/habilitations/dgfip/api_impots_particuliers_editeur.feature:18]: Je me demande si ça ne serait pas plus clair, vu la complexité des règles des scopes, de faire carrément un fichier à part pour déporter tout ça ? Comme ça on refactor les N steps similaires dans le `Contexte`.

File: app/models/authorization_request/api_impot_particulier.rb (hunk 2 above)
Comment (skelz0r) [reply app/models/authorization_request/api_impot_particulier.rb:66]: Par contre j'aimerais avoir un meilleur nom que `validate :validate_*` => `validate :scopes_compatibility` ?

**Review by skelz0r — CHANGES_REQUESTED** [review]
Globalement le code sur les scopes c'est OK, je request des changes pour tous les petits relicats de je ne sais pas quoi qui traîne qui n'ont pas lieu d'être dans cette PR: les `fill_attributes`, les `safety_certification_*`, la factory.

Y'a un peu de naming sur les 3 méthoes de validation aussi où c'est systématiquement `validate :validate_*` ce qui est redondant et peu explicite.

**Comment by skelz0r** [comment]
> Le Front va être géré dans une autre PR

My 2cents là dessus: je pense que ça sera très compliqué de faire un front propre ici, surtout qu'en réalité en voyant les validations sur les scopes (tout du moins les 2 premiers, le 3e je ne le comprends pas), l'UI des checkbox est presque la pire façon de présenter ça à un usager, et imo on gagnerait à shipper ce qu'on a déjà d'un point de vue UX (même si c'est loin d'être idéal mais iso v1), faire la migration, et voir avec la DGFIP pour revoir toute cette logique de scope qui est assez tordu.

**Comment by Isalafont** [comment]
Les safety_certification sont du au rebase avec develop que j'ai fait. Vu que l'on travaillait sur le même fichier spec et features, j'ai rebase avec develop pour gérer les conflicts.

File: spec/factories/authorization_requests.rb [hunk 6]
```diff
@@ -384,7 +389,6 @@
       with_basic_infos
       with_cadre_juridique
       with_personal_data
-      with_cadre_juridique
```
[the remark this answers is not in the document]
Comment (skelz0r) [reply spec/factories/authorization_requests.rb:387]: ah oui en effet. ça aurait mérité un commit à part parce que là on ne comprend pas pourquoi

File: app/models/authorization_request/api_impot_particulier.rb [hunk 7]
```diff
@@ -13,6 +13,32 @@ class AuthorizationRequest::APIImpotParticulier < AuthorizationRequest
...
+      dgfip_indiIFI dgfip_RevDecl_Cat1_Tspr dgfip_RevDecl_Cat5_NonSal
+      dgfip_RevNets_Cat1_Tspr dgfip_RevNets_Cat1_RentOn dgfip_RevNets_Cat2_Rcm
+      dgfip_RevNets_Cat3_PMV dgfip_RevNets_Cat4_Ref dgfip_RevNets_Cat5_NonSal
+      dgfip_PaDeduc_EnfMaj dgfip_PaDeduc_Autres dgfip_EpargRetrDeduc dgfip_IndLep
+    ]
+  ].freeze
+
+  validate :validate_revenue_years_selection, if: -> { current_build_step == 'scopes' }
+  validate :validate_exclusive_years_scope_combination, if: -> { current_build_step == 'scopes' }
+  validate :validate_incompatible_scopes, if: -> { current_build_step == 'scopes' }
```
[the remark this answers is not in the document]
Comment (Isalafont) [reply app/models/authorization_request/api_impot_particulier.rb:40]: Pour une raison que je ne saurais expliquer, sans `if: -> { current_build_step == 'scopes' }`, nous nous trouvons bloqués par la validation à la création de la demande dès la première page car on nous demande de valider les scopes mais ceux ci se trouvent plusieurs steps plus loin. 
D'ailleurs si on retire cette partie la, en laissant cucumber sur le fichier, les tests s'arrêtent rapidement avec pour message d'erreur: 
`Capybara::ElementNotFound: Unable to find field "Destinataire des données" that is not disabled`

[the remark this answers is not in the document]
Comment (skelz0r) [reply app/models/authorization_request/api_impot_particulier.rb:40]: > Pour une raison que je ne saurais expliquer, sans `if: -> { current_build_step == 'scopes' }`, nous nous trouvons bloqués par la validation à la création de la demande dès la première page car on nous demande de valider les scopes mais ceux ci se trouvent plusieurs steps plus loin.

Ok donc c'est bien ce que je propose d'utiliser, le `->(record) { record.need_complete_validation?(:scopes) }` qui fait exactement ton test mais gère tous les edge cases liés au wizard/single page etc etc..

[the remark this answers is not in the document]
Comment (skelz0r) [reply app/models/authorization_request/api_impot_particulier.rb:40]: (et en fait la raison est simple: sur un wizard, il faut bien créer le modèle à la première étape, mais si tu mets des validations sur des fields plus loin il faut bien les ignorer sur les étapes précédentes).

C'est ce cas là https://github.com/zombocom/wicked/wiki/Building-Partial-Objects-Step-by-Step

File: spec/models/authorization_request/api_impot_particulier_spec.rb [hunk 8]
```diff
@@ -1,7 +1,13 @@
 RSpec.describe AuthorizationRequest::APIImpotParticulier, type: :model do
-  describe 'volumetrie validation' do
-    subject(:authorization_request) { build(:authorization_request, :api_impot_particulier_editeur, fill_all_attributes: true, volumetrie_appels_par_minute:, volumetrie_justification:) }
+  subject(:authorization_request) { build(:authorization_request, :api_impot_particulier_editeur, fill_all_attributes: true, scopes:, volumetrie_appels_par_minute:, volumetrie_justification:, safety_certification_begin_date:, safety_certification_end_date:) }
+
+  let(:scopes) { [] }
+  let(:volumetrie_appels_par_minute) { nil }
+  let(:volumetrie_justification) { nil }
+  let(:safety_certification_begin_date) { nil }
+  let(:safety_certification_end_date) { nil }
```
[the remark this answers is not in the document]
Comment (skelz0r) [reply spec/models/authorization_request/api_impot_particulier_spec.rb:8]: Oui c'est évident vu que c'est référencé dans le subject, mais je ne comprends pas pourquoi ils sont ici dans le sens l'intérêt.

File: app/models/authorization_request/api_impot_particulier.rb (hunk 2 above)
Comment (Isalafont) [reply app/models/authorization_request/api_impot_particulier.rb]: A dire vrai, je sèche un peu sur comment renommer cette méthode d'incompatibilité de scopes.

Car on a 2 cas qui se distinguent, un premier array de scope on va dire "speciaux" concernant la demande d'année de revenue (cas ou on veut souhaite avoir l'Avant-dernière année de revenu, si la dernière année de revenu est indisponible et ou les Données fiscales au 31/12 en cas de décès d'un contribuable marié ou pacsé) 
et les scopes incompatibles qui peuvent être de la catégorie "agrégat fiscaux, revenus catégoriels déclarés ainsi que tous les scopes se trouvant dans la catégorie : revenus net, charges déductibles et éligibilité LEP)

File: spec/models/authorization_request/api_impot_particulier_spec.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (Isalafont) [reply spec/models/authorization_request/api_impot_particulier_spec.rb:8]: Je les ai tous placé ici car j'ai besoin de les déclarer pour tous les tests à venir  en même temps que l'on appelle le subject.
ça m'évite de devoir rajouter les valeurs manquantes dans tous les tests.

[the remark this answers is not in the document]
Comment (skelz0r) [reply spec/models/authorization_request/api_impot_particulier_spec.rb:8]: Ok tu as tout remonté d'un test plus bas got it.

File: app/models/authorization_request/api_impot_particulier.rb (hunk 2 above)
Comment (skelz0r) [reply app/models/authorization_request/api_impot_particulier.rb]: Comme dit hier en réunion ce n'est pas grave, faudrait challenger/demander à la DGFIP comment ils nomment cette règle métier.

Tu peux juste renommer en `vallidate :scope_compatibility` (jarter le `validate_`)

File: app/models/authorization_request/api_impot_particulier.rb [hunk 9]
```diff
@@ -13,6 +13,32 @@ class AuthorizationRequest::APIImpotParticulier < AuthorizationRequest
...
+    dgfip_annee_n_moins_2
+    dgfip_annee_n_moins_3
+    dgfip_annee_n_moins_2_si_indispo_n_moins_1
+  ].freeze
+
+  EXCLUSIVE_REVENUE_YEARS = [
+    %w[dgfip_annee_n_moins_2_si_indispo_n_moins_1],
+    %w[dgfip_annee_n_moins_1 dgfip_annee_n_moins_2 dgfip_annee_n_moins_3]
+  ].freeze
+
```
Comment (skelz0r) [inline app/models/authorization_request/api_impot_particulier.rb:27]:
```suggestion
```

Ce n'est plus utilisé.

File: spec/factories/authorization_requests.rb [hunk 10]
```diff
@@ -199,7 +199,7 @@
     trait :with_scopes do
       after(:build) do |authorization_request, evaluator|
         if authorization_request.need_complete_validation? || evaluator.fill_all_attributes
-          next if authorization_request.scopes.any?
+          next if authorization_request.scopes.present?
```
Comment (skelz0r) [inline spec/factories/authorization_requests.rb:202]: imo tu peux revert ici (`present?` n'est pas assez permissif), surtout que ça ne touche plus du tout ta PR

```suggestion
          next if authorization_request.scopes.any?
```

File: spec/factories/authorization_requests.rb [hunk 11]
```diff
@@ -380,9 +380,14 @@
     trait :api_impot_particulier_editeur do
       type { 'AuthorizationRequest::APIImpotParticulier' }
 
+      after(:build) do |authorization_request, _evaluator|
+        next if authorization_request.scopes.any?
+
+        authorization_request.scopes << 'dgfip_annee_n_moins_1' unless AuthorizationRequest::APIImpotParticulier::MANDATORY_REVENUE_YEARS.intersect?(authorization_request.scopes)
```
Comment (skelz0r) [inline spec/factories/authorization_requests.rb:386]:
```suggestion
        authorization_request.scopes << 'dgfip_annee_n_moins_1'
```

le `unless` est tout le temps faux car scopes est systématiquement vide

**Review by skelz0r — CHANGES_REQUESTED** [review]
c'est du mega détail, dès que c'est green tu peux merge sans re-review

File: spec/factories/authorization_requests.rb (hunk 10 above)
Comment (Isalafont) [reply spec/factories/authorization_requests.rb]: Ah non c'est bon, j'avais pas lu le bon bout de code !
